src/esp/gfx/Renderer.cpp

```diff
@@ -39,17 +50,247 @@ namespace Mn = Magnum;
+      context_->makeCurrentPlatform();
+      Mn::GL::Context::makeCurrent(threadContext_);
+      threadOwnsContext_ = true;
+    }
+
+    std::vector<std::vector<RenderCamera::DrawableTransforms>> jobTransforms(
+        jobs_.size());
+
+    for (int i = 0; i < jobs_.size(); ++i) {
+      auto& job = jobs_[i];
+      sensor::VisualSensor& sensor = std::get<0>(job);
+      scene::SceneGraph& sg = std::get<1>(job);
+      RenderCamera::Flags flags = std::get<3>(job);
+
+      auto* camera = sensor.getRenderCamera();
+      jobTransforms[i].reserve(sg.getDrawableGroups().size());
+      for (auto& it : sg.getDrawableGroups()) {
+        it.second.prepareForDraw(*camera);
+        auto transforms = camera->drawableTransformations(it.second);
+        camera->filterTransforms(transforms, flags);
+
+        jobTransforms[i].emplace_back(std::move(transforms));
+      }
+    }
+
+    sgLock_.store(0, std::memory_order_release);
+    std::atomic_notify_all(&sgLock_);
+
+    for (int i = 0; i < jobs_.size(); ++i) {
+      auto& job = jobs_[i];
+      sensor::VisualSensor& sensor = std::get<0>(job);
+      RenderCamera::Flags flags = std::get<3>(job);
+
+      if (!(flags & RenderCamera::Flag::ObjectsOnly))
+        sensor.renderTarget().renderEnter();
+
+      auto* camera = sensor.getRenderCamera();
+      for (auto& transforms : jobTransforms[i]) {
+        camera->draw(transforms, flags);
+      }
+
+      if (!(flags & RenderCamera::Flag::ObjectsOnly))
+        sensor.renderTarget().renderExit();
+    }
+
+    for (auto& job : jobs_) {
+      sensor::VisualSensor& sensor = std::get<0>(job);
+      const Mn::MutableImageView2D& view = std::get<2>(job);
+      RenderCamera::Flags flags = std::get<3>(job);
+      if (flags & RenderCamera::Flag::ObjectsOnly)
+        continue;
+
+      auto sensorType = sensor.specification()->sensorType;
+      if (sensorType == sensor::SensorType::Color)
+        sensor.renderTarget().readFrameRgba(view);
+
+      if (sensorType == sensor::SensorType::Depth)
+        sensor.renderTarget().readFrameDepth(view);
+
+      if (sensorType == sensor::SensorType::Semantic)
+        sensor.renderTarget().readFrameObjectId(view);
+    }
+
+    int jobsDone = jobs_.size();
+    jobs_.clear();
+    return jobsDone;
+  }
+
+  void threadReleaseContext() {
+    if (threadOwnsContext_) {
+      Mn::GL::Context::makeCurrent(nullptr);
+      context_->releasePlatform();
+      threadOwnsContext_ = false;
+    }
+  }
+
+  enum Task : unsigned int { Exit = 0, ReleaseContext = 1, Render = 2 };
+
+  void run() {
+    context_->makeCurrentPlatform();
+    threadContext_ = new Mn::Platform::GLContext{Mn::NoCreate};
+    if (!threadContext_->tryCreate())
+      Mn::Fatal{} << "BackgroundRenderThread: Failed to create OpenGL context";
+
+    Mn::GL::Context::makeCurrent(threadContext_);
+    threadOwnsContext_ = true;
+
+    Mn::GL::Renderer::enable(Mn::GL::Renderer::Feature::DepthTest);
```

Comment:
Is this bit of render-state logic copy-pasted from `Renderer::Impl` constructor? I'm worried about maintaining this. Can we avoid the code duplication?